internal/datasource/git.go

```diff
@@ -291,49 +291,87 @@ func (s *GitSource) Changes(
 		URLs: []string{source.Git.Url},
 	})
 
+	// Determine our target ref. If no Ref is specified by the user,
+	// we default to HEAD which points to whatever the default branch is.
+	targetRef := source.Git.Ref
+	if targetRef == "" {
+		targetRef = "HEAD"
+	}
+
 	// List our refs, equivalent to git ls-remote
 	refs, err := remote.List(&git.ListOptions{Auth: auth})
 	if err != nil {
 		return nil, err
 	}
 
-	// Find the target for HEAD
-	var headTarget plumbing.ReferenceName
+	// Build a map of our refs since we may have to do many lookups
+	refMap := map[plumbing.ReferenceName]*plumbing.Reference{}
 	for _, ref := range refs {
-		if ref.Type() == plumbing.SymbolicReference && ref.Name() == "HEAD" {
-			headTarget = ref.Target()
-			break
+		// We should never get a duplicate ref, but if we do, let's log it
+		// because this will PROBABLY result in some weird behavior.
+		if _, ok := refMap[ref.Name()]; ok {
+			log.Warn("duplicate ref in ls-remote, this shouldn't happen",
+				"name", ref.Name())
 		}
+
+		refMap[ref.Name()] = ref
 	}
-	if headTarget == "" {
-		return nil, status.Errorf(codes.Internal, "HEAD symbolic ref not found")
+
+	// tryRefs is the list of refs we will try to find in the ref list.
+	// The first ref found that matches will be returned first. This contains
+	// all the formats a user may enter for a ref. The order at the time
+	// of writing is the following, where %s is replaced with the user-supplied
+	// ref. This is the same logic `git rev-parse` uses.
+	//
+	//  - "%s",
+	//  - "refs/%s",
+	//  - "refs/tags/%s",
+	//  - "refs/heads/%s",
+	//  - "refs/remotes/%s",
+	//  - "refs/remotes/%s/HEAD",
+	//
+	tryRefs := []string{targetRef}
+	for _, rule := range plumbing.RefRevParseRules {
+		tryRefs = append(tryRefs, fmt.Sprintf(rule, targetRef))
 	}
+	var foundRef *plumbing.Reference
+	for _, tryRef := range tryRefs {
+		ref, ok := refMap[plumbing.ReferenceName(tryRef)]
+		if !ok {
+			continue
+		}
 
-	// Get the actual hash for HEAD
-	var headRef *plumbing.Reference
-	for _, ref := range refs {
-		if ref.Name() == headTarget {
-			headRef = ref
-			break
+		// If the ref is a symbolic reference, we dereference until we find
+		// the target. An example here is HEAD may point to refs/heads/main,
+		// and so on.
+		for ref != nil && ref.Type() == plumbing.SymbolicReference {
+			ref = refMap[ref.Target()]
+		}
```

Comment:
It's unlikely, but if someone wanted to crash a waypoint server, they could configure a symbolic ref loop here and we'd be stuck forever. Maybe just throw the equiv of unix's symlink depth check in here.